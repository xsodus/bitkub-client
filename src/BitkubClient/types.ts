export const BITKUB_API_KEY_HEADER_NAME = "x-btk-apikey";
export const CONTENT_TYPE_HEADER_NAME = "content-type";

export enum BitkubErrorCode {
  NO_ERROR = 0,
  EMPTY_WALLET = 17,
  LOW_AMOUNT = 15,
  // TODO : Add more codes
  // Ref: https://github.com/bitkub/bitkub-official-api-docs/blob/master/restful-api.md#error-codes
}

export enum BitkubSymbolEnum {
  THB_KUB = "THB_KUB",
  THB_SAND = "THB_SAND",
  THB_BTC = "THB_BTC",
  THB_IOST = "THB_IOST",
  //TODO : Add more symbols
}

type BalanceType = {
  available: number;
  reserved: number;
};

export type BitkubBidReturnType = {
  error: BitkubErrorCode;
  result: Array<any>;
};

export type BitkubHeaderType = {
  [BITKUB_API_KEY_HEADER_NAME]: string;
  accept: string;
  [CONTENT_TYPE_HEADER_NAME]: string;
};

export type BitkubSymbolInfo = {
  id: number;
  info: string;
  symbol: BitkubSymbolEnum;
};

export type BitkubSymbolReturnType = {
  error: BitkubErrorCode;
  result: BitkubSymbolInfo[];
};

export type BitkubBalancesReturnType = {
  error: BitkubErrorCode;
  result: Record<string, BalanceType>;
};

export enum BitkubOrderType {
  LIMIT = "limit",
  MARKET = "market",
}

export enum BitkubEnvironment {
  TEST = "test",
  PRODUCTION = "production",
}

export enum BitkubOrderTypeEnum {
  BUY = "buy",
  SELL = "sell",
}

export type BitkubOrderInfoType = {
  action: string;
  symbol: BitkubSymbolEnum;
};
