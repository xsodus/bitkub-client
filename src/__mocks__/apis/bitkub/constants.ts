export const TEST_API_URL = "http://localhost:9876";
