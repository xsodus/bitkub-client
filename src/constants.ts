export const SECURE_API_URL = "https://api.bitkub.com/api";
export const BITKUB_API_KEY_HEADER_NAME = "X-BTK-APIKEY";
export const BITKUB_SIGNATURE_HEADER_NAME = "X-BTK-SIGN";
export const BITKUB_TIMESTAMP_HEADER_NAME = "X-BTK-TIMESTAMP";
export const CONTENT_TYPE_HEADER_NAME = "content-type";
