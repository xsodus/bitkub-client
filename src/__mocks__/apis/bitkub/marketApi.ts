import nock from "nock";
import { TEST_API_URL } from "./constants";

import bitkubBalanceResponse from "./reponses/bitkubBalanceResponse.json";
import bitkubPlaceBidResponse from "./reponses/bitkubBuyResponse.json";
import bitkubCancelOrderResponse from "./reponses/bitkubCancelOrderResponse.json";
import bitkubPlaceAskResponse from "./reponses/bitkubSellResponse.json";
import bitkubTestPlaceBidResponse from "./reponses/bitkubTestBuyResponse.json";

export const createApi = () => {
  nock(TEST_API_URL)
    .post("/v3/market/place-ask")
    .reply(200, bitkubPlaceAskResponse);
  nock(TEST_API_URL).post("/v3/market/balances").reply(200, bitkubBalanceResponse);
  nock(TEST_API_URL)
    .post("/v3/market/place-bid")
    .reply(200, bitkubPlaceBidResponse);
  nock(TEST_API_URL)
    .post("/market/place-bid/test")
    .reply(200, bitkubTestPlaceBidResponse);
  nock(TEST_API_URL)
    .post("/market/place-ask/test")
    .reply(200, bitkubTestPlaceBidResponse);
  nock(TEST_API_URL)
    .post("/v3/market/cancel-order")
    .reply(200, bitkubCancelOrderResponse);
};
