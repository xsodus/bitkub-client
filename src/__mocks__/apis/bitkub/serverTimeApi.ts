import nock from "nock";
import { TEST_API_URL } from "./constants";

export const createApi = (fixedTimestamp = "1705229506193") => {
  nock(TEST_API_URL).persist().get("/v3/servertime").reply(200, fixedTimestamp);
};
