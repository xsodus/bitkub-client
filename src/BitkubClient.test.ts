import nock from "nock";
import BitkubClient from "./BitkubClient";
import { TEST_API_URL } from "./__mocks__/apis/bitkub/constants";
import { createApi as createMarketApi } from "./__mocks__/apis/bitkub/marketApi";
import { createApi as createServerTimeApi } from "./__mocks__/apis/bitkub/serverTimeApi";
import {
  BITKUB_API_KEY_HEADER_NAME,
  BITKUB_SIGNATURE_HEADER_NAME,
  BITKUB_TIMESTAMP_HEADER_NAME,
  CONTENT_TYPE_HEADER_NAME,
} from "./constants";
import {
  BitkubEnvironment,
  BitkubErrorCode,
  BitkubHeaderType,
  BitkubOrderType,
  OrderSide,
  SymbolResponse,
} from "./models";

let client: BitkubClient;

jest.setTimeout(60000);

describe("Bitkub client", () => {
  beforeEach(() => {
    client = new BitkubClient(
      process.env.BITKUB_API_KEY || "test-key",
      process.env.BITKUB_API_SECRET || "test-secret",
      BitkubEnvironment.PRODUCTION
    );

    // Server Time API
    createServerTimeApi();

    // Market API
    createMarketApi();
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it("Get server time", async () => {
    const currentServerTime = await client.getServerTime();
    expect(currentServerTime.data).not.toBeNull();
  });

  it("Can change the api key", async () => {
    const newApiKey = "new-api-key";

    client.apiKey = newApiKey;

    expect(client.apiKey).toMatchSnapshot();
  });

  it("Can generate request header for GET method", async () =>{
    client.baseApiUrl = TEST_API_URL;

    const requestHeaders = await client.buildRequestHeaders('GET','/v3/hello?message=true', { test: "111" });

    expect(requestHeaders).toMatchSnapshot();
  })

  it("Can change the api secret", async () => {
    client.baseApiUrl = TEST_API_URL;
    const oldRequestHeaders = await client.buildRequestHeaders('POST','/v3/hello', { test: "111" });

    client.apiSecret = "XXXXXXXXXXXXXX";

    const newRequestHeaders = await client.buildRequestHeaders('POST','/v3/hello',{ test: "111" });

    expect(oldRequestHeaders).not.toEqual(newRequestHeaders);
  });

  it("Build a request payload including signature key", async () => {
    client.setEnvironment(BitkubEnvironment.PRODUCTION);
    client.baseApiUrl = TEST_API_URL;
    const payload = { id: "1234" };
    const result = await client.buildRequestHeaders('POST','/v3/hello',payload);
    expect(result).toMatchSnapshot();
  });

  it("Get symbols", async () => {
    const result = await client.getSymbols();
    const content: SymbolResponse = result.data;

    console.log("Bitkub All Symbols:", content);
    expect(content).toMatchSnapshot();
  });

  it("Get balances", async () => {
    client.baseApiUrl = TEST_API_URL;

    const result = await client.getBalances();
    const content = result.data;
    expect(content).toMatchSnapshot();
  });

  it("Get bids", async () => {
    const response = await client.getBids("THB_BTC");
    expect(response.data).toHaveProperty("error", BitkubErrorCode.NO_ERROR);

    const [orderId, timestamp, volume, rate, amount] = response.data.result[0];

    expect(orderId).not.toBeUndefined();
    expect(timestamp).not.toBeUndefined();
    expect(volume).not.toBeUndefined();
    expect(rate).not.toBeUndefined();
    expect(amount).not.toBeUndefined();
  });

  it("Get asks", async () => {
    const response = await client.getAsks("THB_BTC");
    expect(response.data).toHaveProperty("error", BitkubErrorCode.NO_ERROR);

    const [orderId, timestamp, volume, rate, amount] = response.data.result[0];

    expect(orderId).not.toBeUndefined();
    expect(timestamp).not.toBeUndefined();
    expect(volume).not.toBeUndefined();
    expect(rate).not.toBeUndefined();
    expect(amount).not.toBeUndefined();
  });

  it("Open position through mocked server", async () => {
    client.setEnvironment(BitkubEnvironment.PRODUCTION);
    client.baseApiUrl = TEST_API_URL;

    const response = await client.placeBid(
      "THB_DOGE",
      100,
      0,
      BitkubOrderType.MARKET,
      "test-client-id"
    );
    expect(response.data).toHaveProperty("error", BitkubErrorCode.NO_ERROR);

    const placeBidResult = response.data.result;

    expect(placeBidResult).toMatchSnapshot();
  });

  it("Close position through mocked server", async () => {
    client.setEnvironment(BitkubEnvironment.PRODUCTION);
    client.baseApiUrl = TEST_API_URL;

    const response = await client.placeAsk(
      "THB_DOGE",
      0.1,
      0,
      BitkubOrderType.MARKET,
      "test-client-id"
    );

    const placeAskResult = response.data.result;

    expect(placeAskResult).toMatchSnapshot();
  });

  it("Open position in sandbox environment with mocked APIs", async () => {
    client = new BitkubClient("", "", BitkubEnvironment.TEST);
    client.baseApiUrl = TEST_API_URL;

    const response = await client.placeBid(
      "THB_BTC",
      100,
      0,
      BitkubOrderType.MARKET
    );
    expect(response.data).toHaveProperty("error", BitkubErrorCode.NO_ERROR);

    const placeBidResult = response.data.result;

    expect(placeBidResult).toMatchSnapshot();
  });

  it("Close position in sandbox environment with mocked APIs", async () => {
    client = new BitkubClient("", "", BitkubEnvironment.TEST);
    client.baseApiUrl = TEST_API_URL;

    const response = await client.placeAsk(
      "THB_KUB",
      0.1,
      0,
      BitkubOrderType.MARKET
    );

    const placeAskResult = response.data.result;

    expect(placeAskResult).toMatchSnapshot();
  });

  it("Place a bid with default order type", async () => {
    client = new BitkubClient("", "", BitkubEnvironment.TEST);
    client.baseApiUrl = TEST_API_URL;

    const response = await client.placeBid("THB_BTC", 100, 0);

    const placeBidResult = response.data.result;

    expect(placeBidResult).toMatchSnapshot();
  });

  it("Ask a bid with default order type", async () => {
    client.setEnvironment(BitkubEnvironment.PRODUCTION);
    client.baseApiUrl = TEST_API_URL;

    const response = await client.placeAsk("THB_DOGE", 0.1, 0);

    const placeAskResult = response.data.result;

    expect(placeAskResult).toMatchSnapshot();
  });

  it("Should cancel an order with hash", async () => {
    client.setEnvironment(BitkubEnvironment.PRODUCTION);
    client.baseApiUrl = TEST_API_URL;

    createServerTimeApi();

    const buyOrderResponse = await client.placeBid(
      "THB_BTC",
      100,
      100000,
      BitkubOrderType.LIMIT
    );

    const { hash } = buyOrderResponse.data.result;

    const response = await client.cancelOrder(hash);

    const cancelOrderResult = response.data.error;

    expect(cancelOrderResult).toBe(0);
  });

  it("Should cancel an order with order id", async () => {
    client.setEnvironment(BitkubEnvironment.PRODUCTION);
    client.baseApiUrl = TEST_API_URL;

    createServerTimeApi();

    const buyOrderResponse = await client.placeBid(
      "THB_BTC",
      100,
      100000,
      BitkubOrderType.LIMIT
    );

    const { id } = buyOrderResponse.data.result;

    const response = await client.cancelOrder(
      null,
      "THB_BTC",
      id,
      OrderSide.BUY
    );

    const cancelOrderResult = response.data.error;

    expect(cancelOrderResult).toBe(0);
  });

  it("Can change the request timeout", async () => {
    client.setEnvironment(BitkubEnvironment.PRODUCTION);
    client.baseApiUrl = TEST_API_URL;

    client.requestTimeout = 1;

    createServerTimeApi();

    await expect(client.getServerTime()).rejects.toThrowError(
      "timeout of 1ms exceeded"
    );
  });
});
