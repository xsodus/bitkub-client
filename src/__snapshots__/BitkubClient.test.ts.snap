// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bitkub client Ask a bid with default order type 1`] = `
{
  "amt": 1,
  "ci": "input_client_id",
  "cre": 37.5,
  "fee": 37.5,
  "hash": "fwQ6dnQWQPs4cbatFGc9LPnpqyu",
  "id": "1",
  "rat": 15000,
  "rec": 15000,
  "ts": 1533834844,
  "typ": "limit",
}
`;

exports[`Bitkub client Build a request payload including signature key 1`] = `
{
  "X-BTK-APIKEY": "test-key",
  "X-BTK-SIGN": "39d6369dd56db3857d97321823565d6f889fb1b2f771b7c1836313e4f571a7e0",
  "X-BTK-TIMESTAMP": 1705229506193,
  "accept": "application/json",
  "content-type": "application/json",
}
`;

exports[`Bitkub client Can change the api key 1`] = `undefined`;

exports[`Bitkub client Can generate request header for GET method 1`] = `
{
  "X-BTK-APIKEY": "test-key",
  "X-BTK-SIGN": "de95614c5e5cdda406c53f11e9816af81e7bfffb7aae88bdc8a1bb351c8c3b3e",
  "X-BTK-TIMESTAMP": 1705229506193,
  "accept": "application/json",
  "content-type": "application/json",
}
`;

exports[`Bitkub client Close position in sandbox environment with mocked APIs 1`] = `
{
  "amt": 100,
  "cre": 0,
  "fee": 0,
  "hash": "test",
  "id": "0",
  "rat": 0,
  "rec": 0,
  "ts": 1642094235,
  "typ": "market",
}
`;

exports[`Bitkub client Close position through mocked server 1`] = `
{
  "amt": 1,
  "ci": "input_client_id",
  "cre": 37.5,
  "fee": 37.5,
  "hash": "fwQ6dnQWQPs4cbatFGc9LPnpqyu",
  "id": "1",
  "rat": 15000,
  "rec": 15000,
  "ts": 1533834844,
  "typ": "limit",
}
`;

exports[`Bitkub client Get balances 1`] = `
{
  "error": 0,
  "result": {
    "1INCH": {
      "available": 0,
      "reserved": 0,
    },
    "AAVE": {
      "available": 0,
      "reserved": 0,
    },
    "ABT": {
      "available": 0,
      "reserved": 0,
    },
    "ADA": {
      "available": 0,
      "reserved": 0,
    },
    "ALGO": {
      "available": 0,
      "reserved": 0,
    },
    "ALPHA": {
      "available": 0,
      "reserved": 0,
    },
    "APE": {
      "available": 0,
      "reserved": 0,
    },
    "ATOM": {
      "available": 0,
      "reserved": 0,
    },
    "AVAX": {
      "available": 0,
      "reserved": 0,
    },
    "AXL": {
      "available": 0,
      "reserved": 0,
    },
    "AXS": {
      "available": 0,
      "reserved": 0,
    },
    "BAL": {
      "available": 0,
      "reserved": 0,
    },
    "BAND": {
      "available": 0,
      "reserved": 0,
    },
    "BAT": {
      "available": 0,
      "reserved": 0,
    },
    "BCH": {
      "available": 0,
      "reserved": 0,
    },
    "BNB": {
      "available": 0.08779727,
      "reserved": 0,
    },
    "BOBA": {
      "available": 0,
      "reserved": 0,
    },
    "BTC": {
      "available": 0.00112343,
      "reserved": 0,
    },
    "BUSD": {
      "available": 0,
      "reserved": 0,
    },
    "CELO": {
      "available": 0,
      "reserved": 0,
    },
    "CHZ": {
      "available": 0,
      "reserved": 0,
    },
    "COMP": {
      "available": 0,
      "reserved": 0,
    },
    "CRV": {
      "available": 0,
      "reserved": 0,
    },
    "CTXC": {
      "available": 0,
      "reserved": 0,
    },
    "CVC": {
      "available": 0,
      "reserved": 0,
    },
    "DAI": {
      "available": 0,
      "reserved": 0,
    },
    "DOGE": {
      "available": 379.58065375,
      "reserved": 0,
    },
    "DON": {
      "available": 0,
      "reserved": 0,
    },
    "DOT": {
      "available": 0,
      "reserved": 0,
    },
    "DYDX": {
      "available": 0,
      "reserved": 0,
    },
    "ENJ": {
      "available": 67.03629032,
      "reserved": 0,
    },
    "ENS": {
      "available": 0,
      "reserved": 0,
    },
    "ETH": {
      "available": 0.01736505,
      "reserved": 0,
    },
    "EXFI": {
      "available": 0,
      "reserved": 0,
    },
    "FLR": {
      "available": 0,
      "reserved": 0,
    },
    "FTM": {
      "available": 0,
      "reserved": 0,
    },
    "FTT": {
      "available": 0,
      "reserved": 0,
    },
    "GAL": {
      "available": 0,
      "reserved": 0,
    },
    "GALA": {
      "available": 0,
      "reserved": 0,
    },
    "GF": {
      "available": 0,
      "reserved": 0,
    },
    "GLM": {
      "available": 0,
      "reserved": 0,
    },
    "GRT": {
      "available": 0,
      "reserved": 0,
    },
    "GT": {
      "available": 0,
      "reserved": 0,
    },
    "HBAR": {
      "available": 0,
      "reserved": 0,
    },
    "ILV": {
      "available": 0,
      "reserved": 0,
    },
    "IMX": {
      "available": 0,
      "reserved": 0,
    },
    "IOST": {
      "available": 0,
      "reserved": 0,
    },
    "JFIN": {
      "available": 0,
      "reserved": 0,
    },
    "KNC": {
      "available": 0,
      "reserved": 0,
    },
    "KSM": {
      "available": 0,
      "reserved": 0,
    },
    "KUB": {
      "available": 0,
      "reserved": 0,
    },
    "LDO": {
      "available": 0,
      "reserved": 0,
    },
    "LINK": {
      "available": 4.15521119,
      "reserved": 0,
    },
    "LRC": {
      "available": 0,
      "reserved": 0,
    },
    "LUNA": {
      "available": 0,
      "reserved": 0,
    },
    "LUNA2": {
      "available": 0,
      "reserved": 0,
    },
    "LYXE": {
      "available": 0,
      "reserved": 0,
    },
    "MANA": {
      "available": 0,
      "reserved": 0,
    },
    "MATIC": {
      "available": 0,
      "reserved": 0,
    },
    "MKR": {
      "available": 0,
      "reserved": 0,
    },
    "NEAR": {
      "available": 0,
      "reserved": 0,
    },
    "OCEAN": {
      "available": 0,
      "reserved": 0,
    },
    "OMG": {
      "available": 0,
      "reserved": 0,
    },
    "OP": {
      "available": 0,
      "reserved": 0,
    },
    "POW": {
      "available": 0,
      "reserved": 0,
    },
    "SAND": {
      "available": 0,
      "reserved": 0,
    },
    "SCRT": {
      "available": 0,
      "reserved": 0,
    },
    "SGB": {
      "available": 0,
      "reserved": 0,
    },
    "SIX": {
      "available": 0,
      "reserved": 0,
    },
    "SNT": {
      "available": 0,
      "reserved": 0,
    },
    "SNX": {
      "available": 0,
      "reserved": 0,
    },
    "SOL": {
      "available": 0,
      "reserved": 0,
    },
    "SOLO": {
      "available": 0,
      "reserved": 0,
    },
    "STG": {
      "available": 0,
      "reserved": 0,
    },
    "SUSHI": {
      "available": 0,
      "reserved": 0,
    },
    "THB": {
      "available": 14780.5,
      "reserved": 0,
    },
    "TRX": {
      "available": 0,
      "reserved": 0,
    },
    "UNI": {
      "available": 0,
      "reserved": 0,
    },
    "USDC": {
      "available": 30.24,
      "reserved": 0,
    },
    "USDT": {
      "available": 0,
      "reserved": 0,
    },
    "WAN": {
      "available": 0,
      "reserved": 0,
    },
    "XLM": {
      "available": 0,
      "reserved": 0,
    },
    "XRP": {
      "available": 0,
      "reserved": 0,
    },
    "XTZ": {
      "available": 0,
      "reserved": 0,
    },
    "YFI": {
      "available": 0,
      "reserved": 0,
    },
    "ZIL": {
      "available": 0,
      "reserved": 0,
    },
    "ZRX": {
      "available": 0,
      "reserved": 0,
    },
  },
}
`;

exports[`Bitkub client Get symbols 1`] = `
{
  "error": 0,
  "result": [
    {
      "id": 1,
      "info": "Thai Baht to Bitcoin",
      "symbol": "THB_BTC",
    },
    {
      "id": 2,
      "info": "Thai Baht to Ethereum",
      "symbol": "THB_ETH",
    },
    {
      "id": 3,
      "info": "Thai Baht to Wancoin",
      "symbol": "THB_WAN",
    },
    {
      "id": 4,
      "info": "Thai Baht to Cardano",
      "symbol": "THB_ADA",
    },
    {
      "id": 5,
      "info": "Thai Baht to OMG Network",
      "symbol": "THB_OMG",
    },
    {
      "id": 6,
      "info": "Thai Baht to Bitcoin Cash",
      "symbol": "THB_BCH",
    },
    {
      "id": 8,
      "info": "Thai Baht to Tether",
      "symbol": "THB_USDT",
    },
    {
      "id": 10,
      "info": "Thai Baht to XRP",
      "symbol": "THB_XRP",
    },
    {
      "id": 12,
      "info": "Thai Baht to Zilliqa",
      "symbol": "THB_ZIL",
    },
    {
      "id": 13,
      "info": "Thai Baht to Status",
      "symbol": "THB_SNT",
    },
    {
      "id": 14,
      "info": "Thai Baht to Civic",
      "symbol": "THB_CVC",
    },
    {
      "id": 15,
      "info": "Thai Baht to Chainlink",
      "symbol": "THB_LINK",
    },
    {
      "id": 17,
      "info": "Thai Baht to IOST",
      "symbol": "THB_IOST",
    },
    {
      "id": 18,
      "info": "Thai Baht to 0x",
      "symbol": "THB_ZRX",
    },
    {
      "id": 19,
      "info": "Thai Baht to Kyber Network",
      "symbol": "THB_KNC",
    },
    {
      "id": 22,
      "info": "Thai Baht to Arcblock",
      "symbol": "THB_ABT",
    },
    {
      "id": 24,
      "info": "Thai Baht to Decentraland",
      "symbol": "THB_MANA",
    },
    {
      "id": 28,
      "info": "Thai Baht to Cortex",
      "symbol": "THB_CTXC",
    },
    {
      "id": 29,
      "info": "Thai Baht to Stellar",
      "symbol": "THB_XLM",
    },
    {
      "id": 30,
      "info": "Thai Baht to SIX",
      "symbol": "THB_SIX",
    },
    {
      "id": 31,
      "info": "Thai Baht to JFIN Coin",
      "symbol": "THB_JFIN",
    },
    {
      "id": 33,
      "info": "Thai Baht to BNB",
      "symbol": "THB_BNB",
    },
    {
      "id": 34,
      "info": "Thai Baht to Powerledger",
      "symbol": "THB_POW",
    },
    {
      "id": 35,
      "info": "Thai Baht to Dogecoin",
      "symbol": "THB_DOGE",
    },
    {
      "id": 37,
      "info": "Thai Baht to TRON",
      "symbol": "THB_TRX",
    },
    {
      "id": 39,
      "info": "Thai Baht to Dai",
      "symbol": "THB_DAI",
    },
    {
      "id": 43,
      "info": "Thai Baht to Tezos",
      "symbol": "THB_XTZ",
    },
    {
      "id": 44,
      "info": "Thai Baht to Algorand",
      "symbol": "THB_ALGO",
    },
    {
      "id": 46,
      "info": "Thai Baht to USD Coin",
      "symbol": "THB_USDC",
    },
    {
      "id": 49,
      "info": "Thai Baht to Basic Attention Token",
      "symbol": "THB_BAT",
    },
    {
      "id": 50,
      "info": "Thai Baht to Maker",
      "symbol": "THB_MKR",
    },
    {
      "id": 56,
      "info": "Thai Baht to Enjin Coin",
      "symbol": "THB_ENJ",
    },
    {
      "id": 65,
      "info": "Thai Baht to Band",
      "symbol": "THB_BAND",
    },
    {
      "id": 66,
      "info": "Thai Baht to Compound",
      "symbol": "THB_COMP",
    },
    {
      "id": 67,
      "info": "Thai Baht to Kusama",
      "symbol": "THB_KSM",
    },
    {
      "id": 68,
      "info": "Thai Baht to Polkadot",
      "symbol": "THB_DOT",
    },
    {
      "id": 69,
      "info": "Thai Baht to NEAR Protocol",
      "symbol": "THB_NEAR",
    },
    {
      "id": 70,
      "info": "Thai Baht to Secret",
      "symbol": "THB_SCRT",
    },
    {
      "id": 71,
      "info": "Thai Baht to Golem",
      "symbol": "THB_GLM",
    },
    {
      "id": 72,
      "info": "Thai Baht to Donnie Finance",
      "symbol": "THB_DON",
    },
    {
      "id": 73,
      "info": "Thai Baht to yearn.finance",
      "symbol": "THB_YFI",
    },
    {
      "id": 74,
      "info": "Thai Baht to Uniswap",
      "symbol": "THB_UNI",
    },
    {
      "id": 75,
      "info": "Thai Baht to Aave",
      "symbol": "THB_AAVE",
    },
    {
      "id": 76,
      "info": "Thai Baht to Stella",
      "symbol": "THB_ALPHA",
    },
    {
      "id": 77,
      "info": "Thai Baht to Ocean Protocol",
      "symbol": "THB_OCEAN",
    },
    {
      "id": 79,
      "info": "Thai Baht to Synthetix",
      "symbol": "THB_SNX",
    },
    {
      "id": 81,
      "info": "Thai Baht to UMA",
      "symbol": "THB_UMA",
    },
    {
      "id": 82,
      "info": "Thai Baht to The Sandbox",
      "symbol": "THB_SAND",
    },
    {
      "id": 83,
      "info": "Thai Baht to Balancer",
      "symbol": "THB_BAL",
    },
    {
      "id": 85,
      "info": "Thai Baht to Curve DAO Token",
      "symbol": "THB_CRV",
    },
    {
      "id": 88,
      "info": "Thai Baht to The Graph",
      "symbol": "THB_GRT",
    },
    {
      "id": 90,
      "info": "Thai Baht to Polygon",
      "symbol": "THB_MATIC",
    },
    {
      "id": 92,
      "info": "Thai Baht to Bitkub Coin",
      "symbol": "THB_KUB",
    },
    {
      "id": 93,
      "info": "Thai Baht to Axie Infinity",
      "symbol": "THB_AXS",
    },
    {
      "id": 95,
      "info": "Thai Baht to Sushi",
      "symbol": "THB_SUSHI",
    },
    {
      "id": 98,
      "info": "Thai Baht to Immutable X",
      "symbol": "THB_IMX",
    },
    {
      "id": 99,
      "info": "Thai Baht to dYdX",
      "symbol": "THB_DYDX",
    },
    {
      "id": 101,
      "info": "Thai Baht to Ethereum Name Service",
      "symbol": "THB_ENS",
    },
    {
      "id": 102,
      "info": "Thai Baht to Boba Token",
      "symbol": "THB_BOBA",
    },
    {
      "id": 103,
      "info": "Thai Baht to Illuvium",
      "symbol": "THB_ILV",
    },
    {
      "id": 104,
      "info": "Thai Baht to Gala",
      "symbol": "THB_GALA",
    },
    {
      "id": 105,
      "info": "Thai Baht to GateToken",
      "symbol": "THB_GT",
    },
    {
      "id": 106,
      "info": "Thai Baht to Songbird",
      "symbol": "THB_SGB",
    },
    {
      "id": 108,
      "info": "Thai Baht to LUKSO",
      "symbol": "THB_LYXE",
    },
    {
      "id": 112,
      "info": "Thai Baht to Chiliz",
      "symbol": "THB_CHZ",
    },
    {
      "id": 114,
      "info": "Thai Baht to GuildFi",
      "symbol": "THB_GF",
    },
    {
      "id": 115,
      "info": "Thai Baht to Coin98",
      "symbol": "THB_C98",
    },
    {
      "id": 117,
      "info": "Thai Baht to Experimental Finance",
      "symbol": "THB_EXFI",
    },
    {
      "id": 118,
      "info": "Thai Baht to Solana",
      "symbol": "THB_SOL",
    },
    {
      "id": 119,
      "info": "Thai Baht to Avalanche",
      "symbol": "THB_AVAX",
    },
    {
      "id": 121,
      "info": "Thai Baht to Fantom",
      "symbol": "THB_FTM",
    },
    {
      "id": 122,
      "info": "Thai Baht to Terra Classic (LUNC)",
      "symbol": "THB_LUNA",
    },
    {
      "id": 123,
      "info": "Thai Baht to ApeCoin",
      "symbol": "THB_APE",
    },
    {
      "id": 124,
      "info": "Thai Baht to Hedera",
      "symbol": "THB_HBAR",
    },
    {
      "id": 126,
      "info": "Thai Baht to Loopring",
      "symbol": "THB_LRC",
    },
    {
      "id": 129,
      "info": "Thai Baht to Celo",
      "symbol": "THB_CELO",
    },
    {
      "id": 131,
      "info": "Thai Baht to Galxe",
      "symbol": "THB_GAL",
    },
    {
      "id": 132,
      "info": "Thai Baht to Sologenic",
      "symbol": "THB_SOLO",
    },
    {
      "id": 133,
      "info": "Thai Baht to Terra",
      "symbol": "THB_LUNA2",
    },
    {
      "id": 134,
      "info": "Thai Baht to Optimism",
      "symbol": "THB_OP",
    },
    {
      "id": 135,
      "info": "Thai Baht to Moonriver",
      "symbol": "THB_MOVR",
    },
    {
      "id": 136,
      "info": "Thai Baht to Glimmer",
      "symbol": "THB_GLMR",
    },
    {
      "id": 139,
      "info": "Thai Baht to 1inch Network",
      "symbol": "THB_1INCH",
    },
    {
      "id": 140,
      "info": "Thai Baht to Cosmos",
      "symbol": "THB_ATOM",
    },
    {
      "id": 141,
      "info": "Thai Baht to LDO",
      "symbol": "THB_LDO",
    },
    {
      "id": 142,
      "info": "Thai Baht to STG",
      "symbol": "THB_STG",
    },
    {
      "id": 143,
      "info": "Thai Baht to AXL",
      "symbol": "THB_AXL",
    },
    {
      "id": 144,
      "info": "Thai Baht to FLOW",
      "symbol": "THB_FLOW",
    },
    {
      "id": 145,
      "info": "Thai Baht to REALX",
      "symbol": "THB_REALX",
    },
    {
      "id": 146,
      "info": "Thai Baht to Flare",
      "symbol": "THB_FLR",
    },
    {
      "id": 147,
      "info": "Thai Baht to Frax Share",
      "symbol": "THB_FXS",
    },
    {
      "id": 148,
      "info": "Thai Baht to Hashflow",
      "symbol": "THB_HFT",
    },
    {
      "id": 149,
      "info": "Thai Baht to Arbitrum",
      "symbol": "THB_ARB",
    },
    {
      "id": 150,
      "info": "Thai Baht to Liquity",
      "symbol": "THB_LQTY",
    },
    {
      "id": 152,
      "info": "Thai Baht to Perpetual Protocol",
      "symbol": "THB_PERP",
    },
    {
      "id": 153,
      "info": "Thai Baht to SPACE ID",
      "symbol": "THB_ID",
    },
    {
      "id": 155,
      "info": "Thai Baht to JOE",
      "symbol": "THB_JOE",
    },
    {
      "id": 156,
      "info": "Thai Baht to Blur",
      "symbol": "THB_BLUR",
    },
    {
      "id": 157,
      "info": "Thai Baht to GNO",
      "symbol": "THB_GNO",
    },
    {
      "id": 158,
      "info": "Thai Baht to WOO",
      "symbol": "THB_WOO",
    },
    {
      "id": 159,
      "info": "Thai Baht to Ankr",
      "symbol": "THB_ANKR",
    },
    {
      "id": 160,
      "info": "Thai Baht to Render Token",
      "symbol": "THB_RNDR",
    },
    {
      "id": 161,
      "info": "Thai Baht to Injective",
      "symbol": "THB_INJ",
    },
    {
      "id": 162,
      "info": "Thai Baht to Worldcoin",
      "symbol": "THB_WLD",
    },
    {
      "id": 164,
      "info": "Thai Baht to Gmx",
      "symbol": "THB_GMX",
    },
    {
      "id": 165,
      "info": "Thai Baht to PLEARN",
      "symbol": "THB_PLN",
    },
    {
      "id": 166,
      "info": "Thai Baht to CyberConnect",
      "symbol": "THB_CYBER",
    },
    {
      "id": 167,
      "info": "Thai Baht to Flux",
      "symbol": "THB_FLUX",
    },
    {
      "id": 168,
      "info": "Thai Baht to Quant",
      "symbol": "THB_QNT",
    },
    {
      "id": 169,
      "info": "Thai Baht to Connext Network",
      "symbol": "THB_NEXT",
    },
    {
      "id": 170,
      "info": "Thai Baht to Celestia",
      "symbol": "THB_TIA",
    },
  ],
}
`;

exports[`Bitkub client Open position in sandbox environment with mocked APIs 1`] = `
{
  "amt": 100,
  "cre": 0,
  "fee": 0,
  "hash": "test",
  "id": "0",
  "rat": 0,
  "rec": 0,
  "ts": 1642094235,
  "typ": "market",
}
`;

exports[`Bitkub client Open position through mocked server 1`] = `
{
  "amt": 1000,
  "ci": "input_client_id",
  "cre": 2.5,
  "fee": 2.5,
  "hash": "fwQ6dnQWQPs4cbatF5Am2xCDP1J",
  "id": "1",
  "rat": 15000,
  "rec": 0.06666666,
  "ts": 1533834547,
  "typ": "limit",
}
`;

exports[`Bitkub client Place a bid with default order type 1`] = `
{
  "amt": 100,
  "cre": 0,
  "fee": 0,
  "hash": "test",
  "id": "0",
  "rat": 0,
  "rec": 0,
  "ts": 1642094235,
  "typ": "market",
}
`;
